src/java.base/share/classes/sun/launcher/LauncherHelper.java

```diff
@@ -840,36 +844,79 @@ private static boolean doesExtendFXApplication(Class<?> mainClass) {
         return false;
     }
 
+    /*
+     * main type flags
+     */
+    private static final int MAIN_WITHOUT_ARGS = 1;
+    private static final int MAIN_NONSTATIC = 2;
+    private static int mainType = 0;
+
+    /*
+     * Return type so that launcher invokes the correct main
+     */
+    public static int getMainType() {
+        return mainType;
+    }
+
+    private static void setMainType(Method mainMethod) {
+        int mods = mainMethod.getModifiers();
+        boolean isStatic = Modifier.isStatic(mods);
+        boolean noArgs = mainMethod.getParameterCount() == 0;
+        mainType = (isStatic ? 0 : MAIN_NONSTATIC) | (noArgs ? MAIN_WITHOUT_ARGS : 0);
+    }
+
     // Check the existence and signature of main and abort if incorrect
-    static void validateMainClass(Class<?> mainClass) {
+    public static void validateMainClass(Class<?> mainClass) {
         Method mainMethod = null;
         try {
-            mainMethod = mainClass.getMethod("main", String[].class);
+            mainMethod = MainMethodFinder.findMainMethod(mainClass);
         } catch (NoSuchMethodException nsme) {
             // invalid main or not FX application, abort with an error
             abort(null, "java.launcher.cls.error4", mainClass.getName(),
                   JAVAFX_APPLICATION_CLASS_NAME);
         } catch (Throwable e) {
             if (mainClass.getModule().isNamed()) {
                 abort(e, "java.launcher.module.error5",
-                      mainClass.getName(), mainClass.getModule().getName(),
-                      e.getClass().getName(), e.getLocalizedMessage());
+                        mainClass.getName(), mainClass.getModule().getName(),
+                        e.getClass().getName(), e.getLocalizedMessage());
             } else {
                 abort(e, "java.launcher.cls.error7", mainClass.getName(),
-                      e.getClass().getName(), e.getLocalizedMessage());
+                        e.getClass().getName(), e.getLocalizedMessage());
             }
         }
 
+        setMainType(mainMethod);
+
         /*
-         * getMethod (above) will choose the correct method, based
+         * findMainMethod (above) will choose the correct method, based
          * on its name and parameter type, however, we still have to
-         * ensure that the method is static and returns a void.
+         * ensure that the method is static (non-preview) and returns a void.
```

Comment:
I think it's probably going to be easier to read and maintain if we moved these checks for `static` and `void` return type into the `MainMethodFinder.findMainMethod(...)` itself.